intelmq/bots/parsers/shadowserver/config.py

```diff
@@ -98,7 +98,22 @@ def convert_int(value):
         return int(value)
 
 
-def convert_host_and_url(value, row):
+def convert_hostname_and_url(value, row):
+    """
+    URLs are split into hostname and path, we can also guess the protocol here.
+    but only guess if the protocol is in a set of known good values.
+    """
+    if row['application'] in ['http', 'https', 'irc']:
+        if row['hostname'] and row['url']:
+            return row['application'] + '://' + row['hostname'] + row['url']
```

Comment:
To be sure we are talking about the same things:

```
       |------------------ Schema-specific Part   ------------------|

 https://max:muster@www.example.com:8080/index.html?p1=A&p2=B#ressource
 \___/   \_/ \____/ \_____________/ \__/\_________/ \_______/ \_______/
   |      |    |           |         |       |          |         |
Scheme    | Password      Host      Port     Path     Query    Fragment
        User
```
Source https://de.wikipedia.org/wiki/Uniform_Resource_Locator


Yes, the Shadowserver `url` field (`row[`url`]) only includes the `path` part of the url.
The full URL can only be determined when the `host` part ist known, maybe the `port` might also be of use.

Guessing the correct `scheme` is tricky here, as this could only be deduced from the `applciation` field and not from other things. But `application` field is `None` in most cases, or contains funky values like `3.390`. So I limited the guessing to http, https and irc, in all other cases simply the `path` part is returned by this function which will cause a failure when the element is added to the event, resulting in a `url` field in the events extra dict.

Damn, that's hard to read...